src/main/java/com/iexec/core/chain/DealWatcherService.java

```diff
@@ -54,69 +55,106 @@ public DealWatcherService(IexecHubService iexecHubService,
+     * a given block to the latest block.
+     * 
+     * @param from start block
+     * @return disposable subscription
+     */
     Disposable subscribeToDealEventFromOneBlockToLatest(BigInteger from) {
         log.info("Watcher DealEvent started [from:{}, to:{}]", from, "latest");
         return iexecHubService.getDealEventObservableToLatest(from)
                 .subscribe(dealEvent -> dealEvent.ifPresent(this::onDealEvent));
     }
 
+    /**
+     * Update last seen block in the database
+     * and run {@link DealEvent} handler.
+     * 
+     * @param dealEvent
+     */
     private void onDealEvent(DealEvent dealEvent) {
         log.info("Received deal [dealId:{}, block:{}]", dealEvent.getChainDealId(), dealEvent.getBlockNumber());
         this.handleDeal(dealEvent.getChainDealId());
-        if (configurationService.getLastSeenBlockWithDeal().intValue() < dealEvent.getBlockNumber().intValue()) {
+        if (configurationService.getLastSeenBlockWithDeal().compareTo(dealEvent.getBlockNumber()) < 0) {
             configurationService.setLastSeenBlockWithDeal(dealEvent.getBlockNumber());
         }
     }
 
+    /**
+     * Handle new onchain deals and add its tasks
+     * to db.
+     * 
+     * @param chainDealId
+     */
     private void handleDeal(String chainDealId) {
-        Optional<ChainDeal> optionalChainDeal = iexecHubService.getChainDeal(chainDealId);
-        if (!optionalChainDeal.isPresent()) {
+        Optional<ChainDeal> oChainDeal = iexecHubService.getChainDeal(chainDealId);
+        if (!oChainDeal.isPresent()) {
+            log.error("Could not get chain deal [chainDealId:{}]", chainDealId);
             return;
         }
-        ChainDeal chainDeal = optionalChainDeal.get();
-
-
-        try {
-            int startBag = chainDeal.getBotFirst().intValue();
-            int endBag = chainDeal.getBotFirst().intValue() + chainDeal.getBotSize().intValue();
-
-            for (int taskIndex = startBag; taskIndex < endBag; taskIndex++) {
-                Optional<Task> optional = taskService.addTask(chainDealId, taskIndex,
-                        BytesUtils.hexStringToAscii(chainDeal.getChainApp().getUri()),
-                        chainDeal.getParams().getIexecArgs(),
-                        chainDeal.getTrust().intValue(),
-                        chainDeal.getChainCategory().getMaxExecutionTime(),
-                        chainDeal.getTag());
-                optional.ifPresent(task -> applicationEventPublisher.publishEvent(new TaskCreatedEvent(task.getChainTaskId())));
-            }
-        } catch (Exception e) {
-            e.printStackTrace();
+        ChainDeal chainDeal = oChainDeal.get();
+        // do not process deals after deadline
+        if (!iexecHubService.isBeforeContributionDeadline(chainDeal)) {
+            log.error("Deal has expired [chainDealId:{}, deadline:{}]",
+                    chainDealId, iexecHubService.getChainDealContributionDeadline(chainDeal));
```

Comment:
Maybe in the future we could have at least 2 iexechub services, one for mutable objects and one for immutable objects (here I see we properly save blockchain calls but iexecHubService.getChainDealContributionDeadline(..) lets think the opposite)

Reply:
Yes, we should get the most out of immutable objects.